shared/src/metabase/shared/formatting/internal/date_formatters.cljc

```diff
@@ -0,0 +1,222 @@
+(ns metabase.shared.formatting.internal.date-formatters
+  "The gory details of transforming date and time styles, with units and other options, into formatting functions.
+
+  This namespace deals with the options only, not with specific dates, and returns reusable formatter functions."
+  (:require
+   [clojure.string :as str]
+   [metabase.shared.formatting.constants :as constants]
+   [metabase.shared.formatting.internal.date-builder :as builder]
+   [metabase.shared.util.log :as log]))
+
+(defn- apply-date-separator [format-list date-separator]
+  (if date-separator
+    (for [fmt format-list]
+      (if (sequential? fmt)
+        [(str/replace (first fmt) #"/" date-separator)]
+        fmt))
+    format-list))
+
+(defn- apply-date-abbreviation [format-list]
+  (for [k format-list]
+    (case k
+      :month-full        :month-short
+      "month-full"       :month-short
+      :day-of-week-full  :day-of-week-short
+      "day-of-week-full" :day-of-week-short
+      k)))
+
+(def ^:private default-date-formats-for-unit
+  "Maps each unit to the default way of formatting that unit.
+  This uses full month and weekday names; abbreviated output replaces these with the short forms later."
+  ;; TODO Do we have (in i18n or utils) helpers for getting localized ordinals?
+  {"year"            [:year]                          ; 2022
+   "quarter"         [["Q"] :quarter [" - "] :year]   ; Q4 - 2022
+   "minute-of-hour"  [:minute-d]                      ; 6, 24
+   "day-of-week"     [:day-of-week-full]              ; Monday; Mon
+   "day-of-month"    [:day-of-month-d]                ; 7, 23
+   "day-of-year"     [:day-of-year]                   ; 1, 24, 365
+   "week-of-year"    [:week-of-year]                  ; CLJS: 1st, 42nd; CLJ: 1, 42 (no ordinals)
+   "month-of-year"   [:month-full]                    ; October; Oct
+   "quarter-of-year" [["Q"] :quarter]})               ; Q4
+
+(def ^:private date-style-to-format-overrides
+  "Map of `{date_style {unit format}}`.
+  If given eg. the style `\"M/D/YYYY\"` but a unit of months, we don't want to use that directly for the format,
+  since it contains days.
+  This map transforms the `date_style + unit` pair to the format data structure."
+  (let [m-y     [:month-d ["/"] :year]
+        mmm-y   [:month-full [", "] :year]]
+    {"M/D/YYYY"           {"month"   m-y}
+     "D/M/YYYY"           {"month"   m-y}
```

Comment:
The cascade is: `style + unit` override, then default for the `unit`, then exactly that `style`.

There's no overrides for `"quarter"` for most of the styles, because the second stage of the fallback has `"quarter" [["Q"] :quarter [" - "] :year]`, which works nicely for `M/D/YYYY`, `D/M/YYYY`, etc.

The only one where it's weird is `YYYY/M/D`, which does have an override to `[:year [" - Q"] :quarter]`.